File: server/routes/analytics.ts
import { Router } from 'express';
import type { QueryKey } from '../../shared/types/dashboard.js';
import { databricksConfigured, getLastSqlError, getLastSqlSuccessAt, resolveMetricView, runAnalyticsQuery, verifyMetricViewAccess } from '../lib/analytics.js';
import { sqlEnvStatus } from '../lib/env.js';
import { consoleDemoMode, sqlColumnSummary } from '../lib/config.js';
import { runningJobCount } from '../lib/jobs.js';
import { describeSummaryProvider } from '../lib/summaryProvider.js';
import { supervisorConfigured } from '../lib/supervisor.js';
import { sqlConfigured, warmupWarehouse } from '../lib/databricksSql.js';
import { getPreloadStatus } from '../lib/preload.js';

const VALID_KEYS: QueryKey[] = [
  'dashboard_dt_kpis',
  'dashboard_dt_period_trend',
  'dashboard_dt_site_by_period',
  'dashboard_dt_category_by_period',
  'dashboard_dt_line_by_period',
  'dashboard_dt_reasons',
  'dashboard_dt_dow',
  'dashboard_dt_top_lines',
  'dashboard_dt_shift_comparison',
  'dashboard_dt_dow_by_shift',
  'dashboard_filter_options',
];

export const analyticsRouter = Router();

analyticsRouter.post('/analytics/query/:queryKey', async (req, res) => {
  const queryKey = req.params.queryKey as QueryKey;
  if (!VALID_KEYS.includes(queryKey)) {
    return res.status(404).json({ error: `Unknown query: ${queryKey}` });
  }

  const params = (req.body?.params || req.body || {}) as Record<string, unknown>;
  const started = Date.now();

  try {
    const result = await runAnalyticsQuery(queryKey, params);
    return res.json({
      query_key: queryKey,
      rows: result.rows,
      row_count: result.rows.length,
      source: result.source,
      _cached: result.cached,
      elapsed_ms: Date.now() - started,
    });
  } catch (e) {
    return res.status(500).json({ error: (e as Error).message });
  }
});

analyticsRouter.get('/warmup', async (_req, res) => {
  try {
    if (sqlConfigured()) {
      await warmupWarehouse();
      return res.json({ ok: true, message: 'SQL warehouse warmed up' });
    }
    return res.json({ ok: true, message: 'SQL not configured' });
  } catch (e) {
    return res.status(500).json({ error: (e as Error).message });
  }
});

analyticsRouter.get('/status', async (_req, res) => {
  const configured = databricksConfigured();
  const env = sqlEnvStatus();
  const metricView = resolveMetricView();
  const sqlTest = configured ? await verifyMetricViewAccess() : { ok: false, error: 'SQL not configured' };

  res.json({
    ok: true,
    architecture: 'sc-manufacturing',
    sql_configured: configured,
    sql_ok: sqlTest.ok,
    sql_test: sqlTest,
    last_sql_error: getLastSqlError(),
    last_sql_success_at: getLastSqlSuccessAt(),
    repo_root: env.repo_root,
    cwd: env.cwd,
    env_file: env.env_file,
    env_search: env.env_search,
    env_vars_set: {
      host: env.host_set,
      token: env.token_set,
      warehouse: env.warehouse_set,
    },
    missing_env: env.missing,
    warehouse: process.env.DATABRICKS_WAREHOUSE_ID || 'NOT SET',
    host: process.env.DATABRICKS_HOST || process.env.DATABRICKS_SERVER_HOSTNAME || 'NOT SET',
    catalog: process.env.DATABRICKS_CATALOG || 'main',
    schema: process.env.DATABRICKS_SCHEMA || '(not set — two-part view name)',
    metric_view: metricView,
    sql_columns: sqlColumnSummary(),
    mode: sqlTest.ok
      ? 'live-sql-metric-view'
      : configured
        ? 'sql-error'
        : consoleDemoMode()
          ? 'demo-fallback'
          : 'production-no-demo',
    demo_mode: consoleDemoMode(),
    ...describeSummaryProvider(),
    supervisor: supervisorConfigured() ? process.env.SUPERVISOR_ENDPOINT_NAME : 'not configured',
    active_supervisor_jobs: runningJobCount(),
    preload: getPreloadStatus(),
  });
});

File: py_server/routes/analytics.py
"""Analytics routes — mirrors server/routes/analytics.ts."""
from __future__ import annotations

import os
import time

from flask import Blueprint, jsonify, request

from py_server.lib.analytics import (
    databricks_configured,
    get_last_sql_error,
    get_last_sql_success_at,
    live_data_required,
    run_analytics_query,
    verify_metric_view_access,
)
from py_server.lib.config import console_demo_mode, resolve_metric_view, sql_column_summary
from py_server.lib.databricks_sql import sql_configured, warmup_warehouse
from py_server.lib.env import sql_env_status
from py_server.lib.jobs import running_job_count
from py_server.lib.preload import get_preload_status
from py_server.lib.summary_provider import describe_summary_provider
from py_server.lib.supervisor import supervisor_configured

VALID_KEYS = [
    'dashboard_dt_kpis',
    'dashboard_dt_period_trend',
    'dashboard_dt_site_by_period',
    'dashboard_dt_category_by_period',
    'dashboard_dt_category_network',
    'dashboard_dt_line_by_period',
    'dashboard_dt_line_network',
    'dashboard_dt_reasons',
    'dashboard_dt_dow',
    'dashboard_dt_top_lines',
    'dashboard_dt_shift_comparison',
    'dashboard_dt_dow_by_shift',
    'dashboard_filter_options',
]

bp = Blueprint('analytics', __name__)


@bp.post('/analytics/query/<query_key>')
def run_query(query_key: str):
    if query_key not in VALID_KEYS:
        return jsonify({'error': f'Unknown query: {query_key}'}), 404

    body = request.get_json(silent=True) or {}
    params = body.get('params') or body or {}
    started = time.time() * 1000

    try:
        result = run_analytics_query(query_key, params)
        rows = result.get('rows', [])
        return jsonify({
            'query_key': query_key,
            'rows': rows,
            'row_count': len(rows),
            'source': result.get('source'),
            '_cached': result.get('cached'),
            'elapsed_ms': int(time.time() * 1000 - started),
        })
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@bp.get('/warmup')
def warmup():
    try:
        if sql_configured():
            warmup_warehouse()
            return jsonify({'ok': True, 'message': 'SQL warehouse warmed up'})
        return jsonify({'ok': True, 'message': 'SQL not configured'})
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@bp.get('/status')
def status():
    configured = databricks_configured()
    env = sql_env_status()
    metric_view = resolve_metric_view()
    sql_test = verify_metric_view_access() if configured else {'ok': False, 'error': 'SQL not configured'}

    if sql_test.get('ok'):
        mode = 'live-sql-metric-view'
    elif configured:
        mode = 'sql-error'
    elif console_demo_mode():
        mode = 'demo-fallback'
    else:
        mode = 'production-no-demo'

    return jsonify({
        'ok': True,
        'architecture': 'sc-manufacturing',
        'sql_configured': configured,
        'sql_ok': sql_test.get('ok'),
        'sql_test': sql_test,
        'last_sql_error': get_last_sql_error(),
        'last_sql_success_at': get_last_sql_success_at(),
        'repo_root': env.get('repo_root'),
        'cwd': env.get('cwd'),
        'env_file': env.get('env_file'),
        'env_search': env.get('env_search'),
        'env_vars_set': {
            'host': env.get('host_set'),
            'token': env.get('token_set'),
            'warehouse': env.get('warehouse_set'),
        },
        'missing_env': env.get('missing'),
        'warehouse': os.getenv('DATABRICKS_WAREHOUSE_ID') or 'NOT SET',
        'host': os.getenv('DATABRICKS_HOST') or os.getenv('DATABRICKS_SERVER_HOSTNAME') or 'NOT SET',
        'catalog': os.getenv('DATABRICKS_CATALOG') or 'main',
        'schema': os.getenv('DATABRICKS_SCHEMA') or '(not set — two-part view name)',
        'metric_view': metric_view,
        'sql_columns': sql_column_summary(),
        'mode': mode,
        'demo_mode': console_demo_mode(),
        'live_data_required': live_data_required(),
        'demo_allowed': console_demo_mode() and not live_data_required(),
        **describe_summary_provider(),
        'supervisor': os.getenv('SUPERVISOR_ENDPOINT_NAME') if supervisor_configured() else 'not configured',
        'active_supervisor_jobs': running_job_count(),
        'preload': get_preload_status(),
    })
